Shaders

atsb opened the pull request:
This adds a very basic GLSL shader that aims to replace the texture combiners currently in the code.  It runs throughout the lifecycle of the renderer, effective nullifying any texture combiners currently in-place.

Thoughts?

atsb committed:
Shaders

atsb commented:
Also proving that I do in-fact, do PR's too lol!

atsb commented:
This is mostly taken from the same structure as the GL4.6 renderer I maintain as part of yamagi.  The structure, style and form are mostly identical.

atsb committed:
Structure fixes

Fixes a typo left over from the transition from YQ2 and moving some stuff around as I forgot it needs the headers..

atsb commented:
<img width="1920" height="1080" alt="sshot019" src="https://github.com/user-attachments/assets/7b7c9e61-ca41-4676-8b87-92f785e150ec" />
<img width="1920" height="1080" alt="sshot015" src="https://github.com/user-attachments/assets/ff1eb712-3993-4374-a51b-01e76d0a2250" />
<img width="1920" height="1080" alt="sshot016" src="https://github.com/user-attachments/assets/48371eba-e424-4ab9-949d-191de269c084" />
<img width="1920" height="1080" alt="sshot017" src="https://github.com/user-attachments/assets/31fef4b8-9e09-4874-a573-6a1b4e0e7eb9" />
<img width="1920" height="1080" alt="sshot018" src="https://github.com/user-attachments/assets/9d518647-12e6-46ca-8d5f-a3cefd41518a" />

The look and feel is very much 'remaster'..  though there's a bug in the annoying final room..  AGAIN!

atsb committed:
Another typo fix

StevenSYS commented:
I've found a bug!
If you press "Apply Settings" in the video menu the lighting resets back to what it previously looked like

Before:
<img width="1920" height="1080" alt="sshot000" src="https://github.com/user-attachments/assets/a6826397-3a7d-4330-a377-f5751fb8ba9c" />

After:
<img width="1920" height="1080" alt="sshot001" src="https://github.com/user-attachments/assets/31c9ab0e-26b7-4fc3-90f9-60e8ccef774f" />

atsb commented:
Aha!  There must be a leak somewhere that pressing it, reverts to the older overbrights and texture combiners..

atsb commented:
Good thing I did a PR and not push straight to master

bubbleguuum commented:
Looks good !

atsb committed:
Fixing the apply settings bug

The shader bind was keeping track of the state of the initialisation struct, even on resets, so I removed that check so that it ALWAYS recompiles on demand.

atsb committed:
Merge branch 'stable' into shaders

atsb commented:
> I've found a bug! If you press "Apply Settings" in the video menu the lighting resets back to what it previously looked like
> 
> Before: <img alt="sshot000" width="1920" height="1080" src="https://private-user-images.githubusercontent.com/139715581/482380936-a6826397-3a7d-4330-a377-f5751fb8ba9c.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTYyNDYxMjgsIm5iZiI6MTc1NjI0NTgyOCwicGF0aCI6Ii8xMzk3MTU1ODEvNDgyMzgwOTM2LWE2ODI2Mzk3LTNhN2QtNDMzMC1hMzc3LWY1NzUxZmI4YmE5Yy5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwODI2JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDgyNlQyMjAzNDhaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0wZDk1NjQ0MmM5N2MwYTMyNmU1MmMwZmViMzlhYmQ5OTg2N2I0NWM2ZGU3OTcyMmFiNGViYzBjNjIyNDBiMmI5JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.1IMWgDcW8SxAm1X3nzEn2SVID4ZcsuH0WV8UgPrsxMY">
> 
> After: <img alt="sshot001" width="1920" height="1080" src="https://private-user-images.githubusercontent.com/139715581/482380980-31c9ab0e-26b7-4fc3-90f9-60e8ccef774f.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTYyNDYxMjgsIm5iZiI6MTc1NjI0NTgyOCwicGF0aCI6Ii8xMzk3MTU1ODEvNDgyMzgwOTgwLTMxYzlhYjBlLTI2YjctNGZjMy05MGY5LTYwZThjY2VmNzc0Zi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwODI2JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDgyNlQyMjAzNDhaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT00YmY5YzQ3ZTg0Yjk5M2I0YmY1ZWYxNWFlNjRlYzlkODgxNjYwZDBmNjYwNTJlM2Q0NzE4MjU0ZTI0M2UwMjlmJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.pDoqNFkeOgBTCjEEGBQcGg8SIJWgdC171vvWn6Qkekg">

FIXED!

atsb commented:
As for the final room, it looks like the texture combiners are doing a two pass in the legacy code and the shaders are only a single pass.  I'll have a look at that function and see what we can do about that.  We will probably need multiple shaders for these things, to really target the individual nature of every individual thing.  Like Glass, transparency, these rooms, etc...

atsb commented:
So I went back to this commit: https://github.com/atsb/Doom64EX-Plus/tree/6495a02aeb1b7ea08d3a9e61b83262c1f3d2f6b6

This is the last 'working' version of the lights.  Something broke and it is somewhere in either i_png, gl_texture, etc..  I've moved over some files to test within our tree and that breaks it, but in the downloaded tree from the commit, it works.  With or without shaders..

atsb committed:
Merge branch 'stable' into shaders

atsb committed:
commenting out shaders

Fixing a few things

atsb committed:
Fixing barrel explosions

boom boom!!

atsb committed:
Merge branch 'stable' into shaders

atsb commented:
I'll just fix everything in here and push it

atsb commented:
I've decided I will rewrite it into a proper skybox and will remove the switch.  Skyboxes only

atsb committed:
Sexy skies!

Skyboxes are back!

atsb committed:
Removing the skybox option

Always on.  It is perfect!

atsb commented:
Skies are the sexiest you've ever seen!

Now will work on the explosions and do something about these shaders

StevenSYS commented:
The skies doesn't work for me now, they just appears as solid black
Also the game keeps spamming the terminal with `D_ShaderInit: Linkage error:\n\n`

atsb commented:
D_ShaderInit: Linkage error:\n\n - this is normal for now.  I've commented out the compiling of the shaders as it kills the texture combiners in the MAP01 final room.

Hmmm...  solid black definitely isn't normal, that's weird..

atsb commented:
<img width="1920" height="1080" alt="sshot011" src="https://github.com/user-attachments/assets/4bdeecc0-0932-40ee-9042-66e0f5f00012" />

So it isn't like this?

atsb committed:
Merge branch 'stable' into shaders

StevenSYS commented:
In Map 30 it looks like this:
<img width="1920" height="1080" alt="MAP30" src="https://github.com/user-attachments/assets/d38fee83-26df-444f-a1ba-02f6bea1e815" />

In Map 28 it looks correct (Minus the texture filtering being linear, when I have it set to nearest):
<img width="1920" height="1080" alt="MAP28" src="https://github.com/user-attachments/assets/1ee7be72-7eb5-4031-9721-c781b37c4bf5" />

And in MAP01 (and the title demo) it's solid black:
<img width="1920" height="1080" alt="MAP01" src="https://github.com/user-attachments/assets/adf616fc-ee87-4c7a-b453-15ec8ddb25b7" />

atsb commented:
I'm going to cry..  I was testing it on the fire skies as they are more visible..

I need a drink

atsb commented:
GL_LINEAR is forced on skies, as they look really terrible with GL_NEAREST

atsb commented:
Oh I KNOW WHY!!!  Oh my gosh..  of course..  Because I was testing it on fire skies, I added the wrap around cases so the sky fits the entire view..  but of course, this will break other skies where the more 'simple sky' option is preferred like in MAP01..

Balls...  I'll write another function for non-fire skies.

atsb commented:
Or actually what I'll do is put back the old code as a new function, lock the view port and force it to display on 'non fire maps' and everything else will use the new skies,,

StevenSYS commented:
> GL_LINEAR is forced on skies, as they look really terrible with GL_NEAREST

I think there should be an option to change the filtering for just the skies, like how there is for the HUD/UI elements

atsb commented:
We can.  Slightly more complex but not impossible

GLint old2DMin, old2DMag;
	glGetTexParameteriv(GL_TEXTURE_2D, GL_TEXTURE_MIN_FILTER, &old2DMin);
	glGetTexParameteriv(GL_TEXTURE_2D, GL_TEXTURE_MAG_FILTER, &old2DMag);
	glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MIN_FILTER, GL_LINEAR);
	glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MAG_FILTER, GL_LINEAR);
#ifdef GL_TEXTURE_CUBE_MAP
	GLint oldCubeMin, oldCubeMag;
	glGetTexParameteriv(GL_TEXTURE_CUBE_MAP, GL_TEXTURE_MIN_FILTER, &oldCubeMin);
	glGetTexParameteriv(GL_TEXTURE_CUBE_MAP, GL_TEXTURE_MAG_FILTER, &oldCubeMag);
	glTexParameteri(GL_TEXTURE_CUBE_MAP, GL_TEXTURE_MIN_FILTER, GL_LINEAR);
	glTexParameteri(GL_TEXTURE_CUBE_MAP, GL_TEXTURE_MAG_FILTER, GL_LINEAR);
#endif

StevenSYS committed:
Added option to change the filtering for skies

StevenSYS commented:
Done!
<img width="1920" height="1080" alt="sshot000" src="https://github.com/user-attachments/assets/e4a64055-02d5-4a02-9839-38dfc9feb9bd" />

Should push my changes to this branch?

atsb commented:
Please do

atsb committed:
3-point filtering

Won't be able to turn this off unless we disable the entire shader system.

Let's leave this on, it's faithful and copies the remaster.

atsb committed:
Merge branch 'shaders' of https://github.com/atsb/Doom64EX-Plus into shaders

atsb commented:
The 3 point filter is in.  This is a hard change, world filtering will need to be removed as it cannot filter on top of the 3-point and we cannot remove it, because removing it, will mean also removing the shader system.  Which is overkill.  The 3-point filter will be staying.  I'll be removing the filtering for the world from the menu but will leave the HUD and SKY untouched (those are exempt from the 3 point filter anyway)

StevenSYS commented:
Turns out this issue is because of a mistake I made in with modification

~~I found another bug~~
~~When you press "Apply Settings" in the video menu it breaks the lighting~~

<img width="1920" height="1080" alt="Lighting" src="https://github.com/user-attachments/assets/b1016a9c-e9f5-46d3-bda4-9b123e5e08ef" />

~~Also the 3-point texture filtering being missing in the image isn't a bug, it's a modification I did to the code that allows you to remove the filtering using a macro~~

atsb commented:
> I found another bug When you press "Apply Settings" in the video menu it breaks the lighting
> 
> <img alt="Lighting" width="1920" height="1080" src="https://private-user-images.githubusercontent.com/139715581/483404292-b1016a9c-e9f5-46d3-bda4-9b123e5e08ef.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTY0NDcxMjksIm5iZiI6MTc1NjQ0NjgyOSwicGF0aCI6Ii8xMzk3MTU1ODEvNDgzNDA0MjkyLWIxMDE2YTljLWU5ZjUtNDZkMy1iZGE0LTliMTIzZTVlMDhlZi5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwODI5JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDgyOVQwNTUzNDlaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0xYjI0OTdiZWQ3NDRjYTQ4NjcxN2NkNDBjMDBmMTUyMzgxN2NlN2JhZTdmZTQxNjcxMGM1NzM3YWNiNGEyNjAyJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.yKl3vmCtcCGWp7JidsWV9ObKeNDVhqtgi_L8c7dIs8c">
> Also the 3-point texture filtering being missing in the image isn't a bug, it's a modification I did to the code that allows you to remove the filtering using a macro

This damn apply settings button..

It is because when you apply settings, it rebakes the GL renderer and doesn't force a shader recompilation..  I'll see what I can do.

And If you want to make anything into a menu option (if you're able to switch between 3-point and normal for example) then go ahead and make those changes.  It would be nice I suppose.  Something like Filtering: 'N64', 'Linear', 'No Filtering'

atsb committed:
Renaming so it is clear what it does

atsb commented:
Ok the skies are fixed I'll add back my skybox and make it look identical to the remaster.

atsb commented:
And fix everything else

atsb committed:
Skies are fixed

Shaders improved too.  As for the 'actual sky'  I think this is a remaster bug where the colours are actually bugged, this was discovered about 2 years ago and obviously can't be fixed on their side.  I found it out when porting over reloaded where marsh would be orange instead of green, because of the values in the SKYDEF.  Let's try to change those values to 'actual known' colours and that'll probably work.  Skies are clearly there, just the wrong colour.

atsb committed:
Merge branch 'stable' into shaders

atsb committed:
Shaders: Improved gating

This protects the fixed function combiners from fighting with the 3-point.  Doesn't fix sectors yet but fixes the 3 point applying to weapons.  Also gates the sky (still not fixed)

atsb committed:
Merge branch 'shaders' of https://github.com/atsb/Doom64EX-Plus into shaders

StevenSYS committed:
Added changing the texturing filtering with 3-point as an option and split the weapon filtering into its own option

StevenSYS commented:
> And If you want to make anything into a menu option (if you're able to switch between 3-point and normal for example) then go ahead and make those changes. It would be nice I suppose. Something like Filtering: 'N64', 'Linear', 'No Filtering'

Done!
I've also split the weapon filtering into its own option

atsb committed:
More shader work

Gating more

atsb committed:
Merge branch 'shaders' of https://github.com/atsb/Doom64EX-Plus into shaders

atsb commented:
Amazing stuff!

atsb commented:
I know it is still a bit buggy, but we should probably merge this in and fix in subsequent branches..  it's going to get messy otherwise

atsb commented:
Ok HERE WE GOOOO

atsb commented:
May the Imps be with us!

StevenSYS commented:
I've already found a bug
The light flashes don't work

atsb commented:
> I've already found a bug The light flashes don't work

Correct.  I've gated all of the GL routines with D_ShaderUnBind and Bind which prevents the shader from being active in those routines, allowing the traditional texture combiner to do what it needs to do..  however I can not find out the reason for the lights no longer working.

StevenSYS commented:
> lights no longer working.

When I said "light flashes", I meant the flashes that happen when you get a berserk, get hurt, etc

atsb commented:
> > lights no longer working.
> 
> When I said "light flashes", I meant the flashes that happen when you get a berserk, damage, etc

Shit...

I'll fix it I guess :D it is likely more gating issues

atsb commented:
This is usually the problem with non-pixel shaders.  You need to juggle two pipelines and prevent any fighting over the world renderer.

atsb commented:
Confirmed as a gating issue..  I commented out the shaders again and everything worked.  So definitely there is still fighting happening.
